compiler/rustc_smir/src/stable_mir/convert/mod.rs

```diff
@@ -0,0 +1,126 @@
+use std::ops::RangeInclusive;
+
+use rustc_smir::Tables;
+use rustc_smir::context::SmirCtxt;
+
+use super::compiler_interface::BridgeTys;
+use crate::rustc_smir;
+
+mod internal;
+mod stable;
+
+impl<'tcx, T> Stable<'tcx> for &T
+where
+    T: Stable<'tcx>,
+{
+    type T = T::T;
+
+    fn stable(
+        &self,
+        tables: &mut Tables<'tcx, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
+    ) -> Self::T {
+        (*self).stable(tables, cx)
+    }
+}
+
+impl<'tcx, T> Stable<'tcx> for Option<T>
+where
+    T: Stable<'tcx>,
+{
+    type T = Option<T::T>;
+
+    fn stable(
+        &self,
+        tables: &mut Tables<'tcx, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
+    ) -> Self::T {
+        self.as_ref().map(|value| value.stable(tables, cx))
+    }
+}
+
+impl<'tcx, T, E> Stable<'tcx> for Result<T, E>
+where
+    T: Stable<'tcx>,
+    E: Stable<'tcx>,
+{
+    type T = Result<T::T, E::T>;
+
+    fn stable(
+        &self,
+        tables: &mut Tables<'tcx, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
+    ) -> Self::T {
+        match self {
+            Ok(val) => Ok(val.stable(tables, cx)),
+            Err(error) => Err(error.stable(tables, cx)),
+        }
+    }
+}
+
+impl<'tcx, T> Stable<'tcx> for &[T]
+where
+    T: Stable<'tcx>,
+{
+    type T = Vec<T::T>;
+    fn stable(
+        &self,
+        tables: &mut Tables<'tcx, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
+    ) -> Self::T {
+        self.iter().map(|e| e.stable(tables, cx)).collect()
+    }
+}
+
+impl<'tcx, T, U> Stable<'tcx> for (T, U)
+where
+    T: Stable<'tcx>,
+    U: Stable<'tcx>,
+{
+    type T = (T::T, U::T);
+    fn stable(
+        &self,
+        tables: &mut Tables<'tcx, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
+    ) -> Self::T {
+        (self.0.stable(tables, cx), self.1.stable(tables, cx))
+    }
+}
+
+impl<'tcx, T> Stable<'tcx> for RangeInclusive<T>
+where
+    T: Stable<'tcx>,
+{
+    type T = RangeInclusive<T::T>;
+    fn stable(
+        &self,
+        tables: &mut Tables<'tcx, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
+    ) -> Self::T {
+        RangeInclusive::new(self.start().stable(tables, cx), self.end().stable(tables, cx))
+    }
+}
+
+/// Trait used to convert between an internal MIR type to a Stable MIR type.
+pub trait Stable<'tcx> {
```

Comment:
hummm... maybe this module should also go inside `stable_mir::internal` or something like `stable_mir::unstable`. We don't provide any stability guarantees on the `Stable` and `Internal` traits.

```suggestion
/// Trait used to convert between an internal MIR type to a Stable MIR type.
///
/// This trait is currently exposed to users so they can have interoperability between internal MIR
/// and StableMIR constructs. However, they should be used seldomly and they have no influence
/// in this crate semver.
#[doc(hidden)]
pub trait Stable<'tcx> {
```

Reply:
Then... how about we bundle all unstable items that have no stability guarantees into a certain module like `stable_mir::unstable`, including `rustc_internal`?